### DISCUSSION TITLE
How to set up 'assembly-clonotypes-by' argument?

### DISCUSSION THREAD

**Discussion by omegahh** [milaboratory/mixcr#1790] [General]
My command for alignment is:

```
mixcr align -f -t 24 -Xmx160g -b default -p generic-amplicon-with-umi --species hsa --rna --tag-pattern "^(UMI:NNNNtNNNNtNNNN)tN{7:8}(R1:*)\^N{17}(R2:*)" --tag-parse-unstranded --rigid-left-alignment-boundary --floating-right-alignment-boundary C --assemble-clonotypes-by '[FR1,CDR1,CDR2_TO_FR4]' --json-report logs/Library.00_S1A08-UQ09-UT09.mixcr_align.json trim_demux/S1A08-UQ09-UT09_R1.fastq.gz trim_demux/S1A08-UQ09-UT09_R2.fastq.gz tmp/S1A08-UQ09-UT09.vdjca 
```

My library is produced by a RACE with UMI protocol designed by myself, and sequenced by PE300 strategy. Thus, in theoretically, it has the full VTranscriptWithP sequence, including 5UTR, L1/L2, and VDJRegion. But there is some gap about in FR2 loci, as shown in the following:

<img width="569" alt="截屏2024-09-12 09 21 13" src="https://github.com/user-attachments/assets/a8941ee0-574c-4767-9543-913cde5c7330">

If I use "VDJRegion," many clones would be discarded, even they still have valid sequences in CDR2_TO_FR4. However, if I use CDR2_TO_FR4, exportClones would lack sequence information for FR1/CDR1.

I tried using a mix-in option like "[FR1,CDR1,CDR2_TO_FR4]" during alignment, but the software indicated that the order was incorrect, and it seems that the software would scramble the order.

<img width="886" alt="截屏2024-09-12 09 30 54" src="https://github.com/user-attachments/assets/8bb8da79-86a3-4299-825f-f950551822db">

I also tried parameters like "{FR1Begin:CDR1End}+{CDR2Begin:FR4End}", but the error messages from the software were not understandable to me.

<img width="1920" alt="截屏2024-09-12 09 37 50" src="https://github.com/user-attachments/assets/78e6fef1-d41d-4cbc-aec1-ca0bb9027cfd">

I would like to know how to handle this situation? I want to make the most of the gene regions that can be covered by the sequencing data, such as FR1+CDR1+CDR2_TO_FR4.

Additionally, I would like to ask if MiXCR supports a flexible strategy: if a clone covers a large area, the assembled region should be sufficiently wide to cover more gene features, but if the coverage area is short, then use fewer gene feature for assembly. For example, I could list possible gene features from complete to partial, like [VDJRegion, FR1+CDR2_TO_FR4, CDR2_TO_FR4], and then the software tries them one by one, which could help preserve as many clones and gene features as possible.

Looking forward to your reply :)

**Comment by mizraelson** [comment]
Hi,
- Is there a reason you don’t use the `mixcr analyze` command?
- You can set the feature to: `--assemble-clonotypes-by [{FR1Begin:CDR1End},{CDR2Begin:FR4End}]`.
- The combination of parameters `--tag-parse-unstranded --tag-pattern "^(UMI:NNNNtNNNNtNNNN)tN{7:8}(R1:*)\^N{17}(R2:*)"` doesn’t quite make sense because there are no anchor points to determine whether 7 or 8 nucleotides should be skipped. The same applies to `--tag-parse-unstranded`, as there is no sequence to determine in which read the UMI is located. It should be, for example:

`^(UMI:NNNNtNNNNtNNNN)tN{7:8}atgggct(R1:*)\^N{17}(R2:*)`

or simply:

`^(UMI:N{14))N{7}(R1:*)\^N{17}(R2:*)` without `--tag-parse-unstranded`, but in that case, you have to be sure the UMI is always in R1.

- Regarding the “flexible strategy,” you can assemble clones by CDR3 during the assemble step preserving the alignments with `-Massemble.clnaOutput=true` and then use `mixcr assembleContigs` to extend the sequence as much as possible.

**Reply by omegahh** [reply]
Thank you for replying. I tried the command you suggested: `--assemble-clonotypes-by [{FR1Begin:CDR1End},{CDR2Begin:FR4End}]`, but still got error about wrong order.

<img width="1457" alt="截屏2024-09-16 14 01 39" src="https://github.com/user-attachments/assets/6d3ee484-ef70-4138-af09-1092aae82fa4">

**Reply by mizraelson** [reply]
This looks like a bug in the latest version. We will fix it as soon as possible. Meanwhile you can try v 4.6.

**Reply by omegahh** [reply]
I'm delighted to hear this news. I'm very sorry for missing this reply.

**Comment by omegahh** [comment]
And about "flexible strategy" for extracting as many features as possible from alignments. You recommend to use "mixcr assembleContigs" command, while I found that this command is not for my case according to the documentations (https://mixcr.com/mixcr/reference/mixcr-assembleContigs/)?

Besides, I tried a lot but all failed for setting "clnaOutput=true", as shown in these screenshots. Can you help me?

<img width="1632" alt="截屏2024-09-22 13 33 20" src="https://github.com/user-attachments/assets/1fa83822-504d-450f-8a85-567872a63efd">
<img width="1920" alt="截屏2024-09-22 13 34 02" src="https://github.com/user-attachments/assets/81d4dfd9-e6f1-46fa-9cb4-dbaacae95361">
<img width="1920" alt="截屏2024-09-22 13 34 47" src="https://github.com/user-attachments/assets/809b4621-4bd3-414a-9187-94180439d34c">
<img width="1920" alt="截屏2024-09-22 13 35 37" src="https://github.com/user-attachments/assets/cb80b97d-1eba-461e-b8eb-bf7b5987b8e7">
<img width="1920" alt="截屏2024-09-22 13 38 29" src="https://github.com/user-attachments/assets/e4138d3b-5c91-4969-85b9-5a4271817519">
<img width="1920" alt="截屏2024-09-22 13 39 23" src="https://github.com/user-attachments/assets/008d0a8f-2bcc-4360-bd9c-6fafd7c07582">

I really cannot find the right command for this argument.

**Comment by omegahh** [comment]
In short, I have PE300 sequenced libraries produced by a RACE with UMI T/BCR sequencing protocol. After alignment, I found that about 50% of reads covers VDJRegion, and others have sequencing gap in FR2. 

So 1). If I 'assemble-clonotypes-by CDR2_TO_FR4', those exported clones will not export "FR1" and "CDR1", which are both covered for sure. 2). If I 'assemble-clonotypes-by VDJRegion', a lot of reads will be dropped. 3). If I use "[{FR1Begin:CDR1End},{CDR2Begin:FR4End}]", the program reports error about wrong order.

Is there anyway of getting as many gene features as possible, as well as clonotypes? I really need your help, Thank you very much! @mizraelson

**Reply by mizraelson** [reply]
Hi, can you try the command bellow and let me know how it works on your end:

```
mixcr analyze generic-amplicon-with-umi \
--species hsa \
--tag-pattern "^(UMI:N{14))N{7}(R1:*)\^N{17}(R2:*)" \
--rna \
--rigid-left-alignment-boundary \
--floating-right-alignment-boundary C \
--assemble-clonotypes-by CDR3 \
--add-step assembleContigs \
inputName_{{R}}.fastq.gz \
output
```

**Reply by omegahh** [reply]
My apologies for the delayed response. Here is the log output from one of my samples after running the command you recommended. The reason for the half alignment rate is because the library preparation kit is directionless for insert DNAs.
https://justpaste.it/gvul6

**Reply by omegahh** [reply]
I also run 'findAlleles' for my TCR & BCR libraries, and run 'findShmTrees' for BCR libraries. Can I set 'assemble-configs-by VDJRegion'?
<img width="513" alt="截屏2024-10-05 07 14 10" src="https://github.com/user-attachments/assets/1572cd43-23cc-4fae-9865-cbc3f9313c53">

<img width="1477" alt="截屏2024-10-05 07 19 16" src="https://github.com/user-attachments/assets/86f18829-1c7d-4e8a-ad4b-147b8b719c02">

**Reply by omegahh** [reply]
@mizraelson can you see the log I pasted?

**Reply by mizraelson** [reply]
Yes, to run `findAlleles` and `findShmTrees`, all clones must be covered equally. This means you can’t build SHM trees from a dataset where some clones have gaps in FR2 while others don’t. This is due to the fundamental issue that it’s unclear where to place clones when they lack information about hypermutations in certain parts of the sequence. You can enable this feature by adding the following parameter to the command, but keep in mind that it will filter out clones that don’t fully cover the VDJregion.

```
mixcr analyze generic-amplicon-with-umi \
--species hsa \
--tag-pattern "^(UMI:N{14))N{7}(R1:*)\^N{17}(R2:*)" \
--rna \
--rigid-left-alignment-boundary \
--floating-right-alignment-boundary C \
--assemble-clonotypes-by CDR3 \
--add-step assembleContigs \
--assemble-contigs-by VDJRegion \
inputName_{{R}}.fastq.gz \
output
```

**Reply by omegahh** [reply]
I see. I believe the software will meet my requirements if it can be run with the argument '--assemble-clonotypes-by [{FR1Begin:CDR1End},{CDR2Begin:FR4End}].' Since the majority of the reads encompass these two regions, it's crucial for my analysis. However, I've encountered an issue with the latest version of MiXCR, which reports an error stating that the regions provided are not in the correct order. Is there a way to resolve this issue?

**Reply by mizraelson** [reply]
I have actually tried the command bellow with the [latest MiXCR version](https://link.milaboratory.com/software/mixcr/mixcr-develop.zip) and it worked:

```
mixcr analyze generic-amplicon-with-umi \
--species hsa \
--tag-pattern "^(UMI:N{14))N{7}(R1:*)\^N{17}(R2:*)" \
--rna \
--rigid-left-alignment-boundary \
--floating-right-alignment-boundary C \
--assemble-clonotypes-by "[{FR1Begin:CDR1End},{CDR2Begin:FR4End}]" \
inputName_{{R}}.fastq.gz \
output
```

Can you please try it on your end?

**Reply by omegahh** [reply]
Ohhh, the 'analyze' command succeed, but run with `align`, `refineTagsAndSort`, `assemble` in a step by step way failed. I prefer the later one because I can run each steps and output corresponding json reports separately. Considering that I have about 1 thousand samples, building up an automatic pipeline and separating time and disk consuming steps is very important. Especially if I still need to optimize arguments for some step. For example, when I adjust arguments for `assemble`, I don't need to run `align` and `refineTagsAndSort` again. And from a pipeline coding perspective, only a config file should be modified, while no code should be added or modified.

**Reply by omegahh** [reply]
I spent a lot of time in reading and conducting codes in `src/main/resources/presets`. I thought I understand how these yaml file manage steps and blocks. But after some tests, I find out that run by `analyze` still has some differences with run by separated steps described in the presetXX.yaml file.

**Reply by mizraelson** [reply]
Hi, the step by step approach should also work. It would be:

```
mixcr align -f \
--report align.report.txt \
--json-report align.report.json \
--preset generic-amplicon-with-umi \
--save-output-file-names align.list.tsv \
--rigid-left-alignment-boundary \
--floating-right-alignment-boundary C \
--rna \
--species hsa \
--tag-pattern "^(UMI:N{14))N{7}(R1:*)\^N{17}(R2:*)" \
--assemble-clonotypes-by "[{FR1Begin:CDR1End},CDR2_TO_FR4]" \
inputName_{{R}}.fastq.gz \
output.alignments.vdjca
```

```
mixcr refineTagsAndSort -f \
--report refine.report.txt \
--json-report refine.report.json \
output.alignments.vdjca \
output.refined.vdjca
```

```
mixcr assemble -f \
--report assemble.report.txt \
--json-report assemble.report.json \
output.refined.vdjca output.clns
```

```
mixcr exportClones -f output.clns routput.clones.tsv
```

The pipeline in analyze is fully describes in a YAML file, but all the inheritances should be taken into consideration.

**Reply by omegahh** [reply]
Exactly, that's what I did. After doing some more tests today, I found that it should be the "assemble" step where I used the "--sort-by-sequence" parameter, which is why I got the "not ordered" error. You can try it on your end.
<img width="1076" alt="截屏2024-10-12 21 41 49" src="https://github.com/user-attachments/assets/93e3169c-ebde-4d17-b7f5-25f8a80fc1a2">

**Reply by mizraelson** [reply]
Why do you use `--sort-by-sequence` ?

**Reply by omegahh** [reply]
Because I want to calculate overlap values among samples, and the help command tells me it. Maybe I misunderstood the meaning of this argument. Never mind, I don't need it anymore.
<img width="972" alt="截屏2024-10-13 14 40 19" src="https://github.com/user-attachments/assets/ac3f48d6-f35d-4848-bb33-692abd899445">

**Reply by mizraelson** [reply]
Yes, there is no need for this parameter. You can still use MiXCR to overlap your repertoires. Let me know if you need help with this.
